File: leetcode-in-python/solution/121_best_time_to_buy_and_sell_stock.py
from typing import List
import unittest


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        assert len(prices) != 0, "prices empty"
        ans = 0
        min_buy_price = prices[0]
        for cur_price in prices:
            if min_buy_price < cur_price:
                ans = max(ans , cur_price - min_buy_price)
            else:
                min_buy_price = cur_price
        return ans


class TestSolution(unittest.TestCase):

    def setUp(self):
        self.test_case = [
            ([7, 1, 5, 3, 6, 4], 5),
            ([7, 6, 4, 3, 1], 0),
            ([5], 0),
            ([4, 5], 1),
        ]
        self.s = Solution()

    def test_solution(self):
        for nums, answer in self.test_case:
            ans = self.s.maxProfit(nums)
            self.assertEqual(answer,ans , (nums, answer))


if __name__ == '__main__':
    unittest.main()

File: leetcode-in-rust/src/solution/_121_best_time_to_buy_and_sell_stock.rs
pub struct Solution {}

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut max_profit = 0;
        let mut min_price = prices[0];
        for cur_price in prices {
            if cur_price < min_price {
                min_price = cur_price;
            } else if max_profit < cur_price - min_price {
                max_profit = cur_price - min_price
            }
        }
        max_profit
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solution() {
        let test_case = vec![
            (vec![7, 1, 5, 3, 6, 4], 5),
            (vec![7, 6, 4, 3, 1], 0),
            (vec![5], 0),
            (vec![4, 5], 1),
        ];
        for (nums, answer) in test_case {
            let ans = Solution::max_profit(nums);
            assert_eq!(ans, answer);
        }
    }
}
